### PR TITLE
[sled-agent] Support bundle storage api

### PR DESCRIPTION
**Pull request by smklein**
This PR implements raw bundle storage for support bundles.

In the future, I expect that Nexus will invoke these APIs to store and later access portions of storage bundles.

- Provides an API to create, list, query, and delete support bundles
- Implements range request support for these APIs

Fixes https://github.com/oxidecomputer/omicron/issues/1599

### CODE REVIEW

File: sled-agent/api/src/lib.rs [hunk 1]
```diff
@@ -581,6 +637,79 @@ pub struct VmmPathParam {
...
+#[derive(Deserialize, JsonSchema)]
+pub struct SupportBundleFilePathParam {
+    #[serde(flatten)]
+    pub parent: SupportBundlePathParam,
+}
+
+/// Metadata about a support bundle
+#[derive(Deserialize, Serialize, JsonSchema)]
+pub struct SupportBundleCreateQueryParams {
+    pub hash: ArtifactHash,
```
Comment (smklein) [inline sled-agent/api/src/lib.rs:673]: TODO: Add the reservation / quota here?

File: range-requests/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+[package]
```
Comment (smklein) [inline range-requests/Cargo.toml:1]: `range-requests` should be reviewed as a part of https://github.com/oxidecomputer/omicron/pull/6963 -- feel free to ignore these files; there should be no major changes.

File: sled-storage/src/manager.rs [hunk 3]
```diff
@@ -955,8 +955,8 @@ impl StorageManager {
             err: None,
         };
 
-        let mountpoint_path =
-            config.name.mountpoint(ZPOOL_MOUNTPOINT_ROOT.into());
+        let mountpoint_root = &self.resources.disks().mount_config().root;
```
Comment (smklein) [inline sled-storage/src/manager.rs:958]: This change is split out as a separate PR here: https://github.com/oxidecomputer/omicron/pull/7110

It's necessary for performing the testing in `sled-agent/src/support-bundle/storage.rs`

File: sled-storage/src/manager_test_harness.rs [hunk 4]
```diff
@@ -83,10 +83,9 @@ pub struct StorageManagerTestHarness {
 impl Drop for StorageManagerTestHarness {
     fn drop(&mut self) {
         if let Some(vdev_dir) = self.vdev_dir.take() {
-            eprintln!(
+            eprint!(
```
Comment (smklein) [inline sled-storage/src/manager_test_harness.rs:86]: This change is split out as a separate PR here: https://github.com/oxidecomputer/omicron/pull/7111

It's just a nice-to-have for incremental testing.

File: sled-agent/src/sled_agent.rs [hunk 5]
```diff
@@ -691,6 +697,11 @@ impl SledAgent {
         .unwrap(); // we retry forever, so this can't fail
     }
 
+    /// Accesses the [SupportBundleManager] API.
+    pub(crate) fn as_support_bundle_storage(&self) -> SupportBundleManager<'_> {
+        SupportBundleManager::new(&self.log, self.storage())
+    }
```
Comment (smklein) [inline sled-agent/src/sled_agent.rs:703]: Rather than directly implementing the support bundle API as part of the Sled Agent, I split it into a different structure to make testing easier.

That's why I'm calling this function in the `http_entrypoints`

File: sled-agent/src/support_bundle/mod.rs [hunk 6]
```diff
@@ -0,0 +1,6 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+pub mod queries;
```
Comment (smklein) [inline sled-agent/src/support_bundle/mod.rs:5]: FYI @papertigers - we both used the `support_bundle.rs` module, so I split your stuff into "queries" and my stuff into "storage". Hopefully the re-organization makes sense?

File: sled-agent/src/support_bundle/storage.rs [hunk 7]
```diff
@@ -0,0 +1,1393 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (smklein) [inline sled-agent/src/support_bundle/storage.rs:1]: This is the meat-and-potatoes of this PR -- I tried to add quite a few tests here. Happy to add more, as requested!

File: sled-storage/src/manager_test_harness.rs (hunk 4 above)
Comment (smklein) [reply sled-storage/src/manager_test_harness.rs]: Merged now; resolving this comment

File: sled-storage/src/manager.rs (hunk 3 above)
Comment (smklein) [reply sled-storage/src/manager.rs]: Merged now; resolving this comment

File: sled-agent/api/src/lib.rs [hunk 8]
```diff
@@ -157,6 +159,60 @@ pub trait SledAgentApi {
...
+    #[endpoint {
+        method = GET,
+        path = "/support-bundles/{zpool_id}/{dataset_id}"
+    }]
+    async fn support_bundle_list(
+        rqctx: RequestContext<Self::Context>,
+        path_params: Path<SupportBundleListPathParam>,
+    ) -> Result<HttpResponseOk<Vec<SupportBundleMetadata>>, HttpError>;
+
+    /// Create a service bundle within a particular dataset
```
Comment (wfchandler) [inline sled-agent/api/src/lib.rs:172]: Confusing to use a name that doesn't match the API path.

```suggestion
    /// Create a support bundle within a particular dataset
```

File: sled-agent/api/src/lib.rs [hunk 9]
```diff
@@ -157,6 +159,60 @@ pub trait SledAgentApi {
...
+        path = "/support-bundles/{zpool_id}/{dataset_id}/{support_bundle_id}"
+    }]
+    async fn support_bundle_create(
+        rqctx: RequestContext<Self::Context>,
+        path_params: Path<SupportBundlePathParam>,
+        query_params: Query<SupportBundleCreateQueryParams>,
+        body: StreamingBody,
+    ) -> Result<HttpResponseCreated<SupportBundleMetadata>, HttpError>;
+
+    /// Fetch a service bundle from a particular dataset
```
Comment (wfchandler) [inline sled-agent/api/src/lib.rs:184]:
```suggestion
    /// Fetch a support bundle from a particular dataset
```

File: sled-agent/api/src/lib.rs [hunk 10]
```diff
@@ -157,6 +159,60 @@ pub trait SledAgentApi {
...
+        method = GET,
+        path = "/support-bundles/{zpool_id}/{dataset_id}/{support_bundle_id}"
+    }]
+    async fn support_bundle_get(
+        rqctx: RequestContext<Self::Context>,
+        path_params: Path<SupportBundlePathParam>,
+        body: TypedBody<SupportBundleGetQueryParams>,
+    ) -> Result<http::Response<Body>, HttpError>;
+
+    /// Fetch a service bundle from a particular dataset
```
Comment (wfchandler) [inline sled-agent/api/src/lib.rs:195]:
```suggestion
    /// Fetch a support bundle from a particular dataset
```

File: sled-agent/api/src/lib.rs [hunk 11]
```diff
@@ -157,6 +159,60 @@ pub trait SledAgentApi {
...
+        method = HEAD,
+        path = "/support-bundles/{zpool_id}/{dataset_id}/{support_bundle_id}"
+    }]
+    async fn support_bundle_head(
+        rqctx: RequestContext<Self::Context>,
+        path_params: Path<SupportBundlePathParam>,
+        body: TypedBody<SupportBundleGetQueryParams>,
+    ) -> Result<http::Response<Body>, HttpError>;
+
+    /// Delete a service bundle from a particular dataset
```
Comment (wfchandler) [inline sled-agent/api/src/lib.rs:206]:
```suggestion
    /// Delete a support bundle from a particular dataset
```

**Review by wfchandler — COMMENTED** [review]
Thanks @smklein, just a few small suggestions from me.

File: range-requests/Cargo.toml (hunk 2 above)
Comment (smklein) [reply range-requests/Cargo.toml:1]: Merged; marking this resolved

File: sled-agent/src/support_bundle/storage.rs [hunk 12]
```diff
@@ -0,0 +1,1393 @@
...
+    // this should call "by_name_seek" instead.
+    let reader = archive.by_name(&entry_path)?;
+
+    let mut reader: Box<dyn std::io::Read> = match range {
+        Some(range) => Box::new(skip_and_limit(
+            reader,
+            range.start() as usize,
+            range.content_length().get() as usize,
+        )?),
+        None => Box::new(reader),
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/support_bundle/storage.rs]: Nice, good catch. Updated.

File: sled-agent/src/support_bundle/storage.rs [hunk 13]
```diff
@@ -0,0 +1,1393 @@
...
+            }
+            Error::Storage(err) => HttpError::from(ExternalError::from(err)),
+            Error::Io(err) => HttpError::for_internal_error(err.to_string()),
+            Error::Range(err) => HttpError::for_internal_error(err.to_string()),
+            Error::Zip(err) => match err {
+                ZipError::FileNotFound => HttpError::for_not_found(
+                    None,
+                    "Entry not found".to_string(),
+                ),
+                err => HttpError::for_internal_error(err.to_string()),
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/support_bundle/storage.rs]: Probably not - I'll update it to use the error chain when converting to strings here.

File: sled-agent/src/support_bundle/storage.rs [hunk 14]
```diff
@@ -0,0 +1,1393 @@
...
+        dataset_id: DatasetUuid,
+    ) -> Result<DatasetConfig, Error> {
+        let datasets_config = self.storage.datasets_config_list().await?;
+        let dataset = datasets_config
+            .datasets
+            .get(&dataset_id)
+            .ok_or_else(|| Error::DatasetNotFound)?;
+
+        if dataset.id != dataset_id || dataset.name.pool().id() != zpool_id {
+            return Err(Error::DatasetNotFound);
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/support_bundle/storage.rs]: `dataset.id != dataset_id` would indicate a bad value from the configuration -- which **shouldn't** happen -- but the value is serialized, so it's **plausible** if the ledger storing this configuration was corrupted.

The pool UUID not matching would be also unlikely; that would be a sign that the *dataset exists* but is on the wrong zpool. We don't currently have a way to move datasets between zpools, but if we did, I suppose that condition would be possible.

I can make explicit error cases for these, to make them more apparent.

File: sled-agent/src/support_bundle/storage.rs [hunk 15]
```diff
@@ -0,0 +1,1393 @@
...
+            };
+
+            let bundle = SupportBundleMetadata { support_bundle_id, state };
+            bundles.push(bundle);
+        }
+
+        Ok(bundles)
+    }
+
+    /// Returns the hex, lowercase sha2 checksum of a file at `path`.
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/support_bundle/storage.rs]: Updated

File: sled-agent/src/support_bundle/storage.rs [hunk 16]
```diff
@@ -0,0 +1,1393 @@
...
+        ));
+        let root =
+            self.get_configured_dataset(zpool_id, dataset_id).await?.name;
+        let dataset =
+            NestedDatasetLocation { path: support_bundle_id.to_string(), root };
+        // The mounted root of the support bundle dataset
+        let support_bundle_dir = dataset
+            .mountpoint(illumos_utils::zpool::ZPOOL_MOUNTPOINT_ROOT.into());
+        let support_bundle_path = support_bundle_dir.join("bundle");
+        let support_bundle_path_tmp = support_bundle_dir.join("bundle.tmp");
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/support_bundle/storage.rs]: I think I hadn't considered this as a problem because the Nexus writing the support bundle should have unique ownership off it -- support bundles are assigned to an individual Nexus for creation and storage -- but regardless, I can use a randomized path.

File: sled-agent/src/support_bundle/storage.rs [hunk 17]
```diff
@@ -0,0 +1,1393 @@
...
+        )
+        .await
+        .expect("Should have created support bundle");
+
+        // The bundle should now appear "Complete"
+        let bundles = mgr.list(harness.zpool_id, dataset_id).await.unwrap();
+        assert_eq!(bundles.len(), 1);
+        assert_eq!(bundles[0].support_bundle_id, support_bundle_id);
+        assert_eq!(bundles[0].state, SupportBundleState::Complete);
+
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/support_bundle/storage.rs:1208]: Sure thing, done

File: sled-agent/src/support_bundle/storage.rs [hunk 18]
```diff
@@ -0,0 +1,1393 @@
...
+            response.headers()[CONTENT_LENGTH],
+            GREET_DATA.len().to_string()
+        );
+        assert_eq!(
+            response.headers()[CONTENT_TYPE],
+            "application/octet-stream"
+        );
+        assert_eq!(response.headers()[ACCEPT_RANGES], "bytes");
+
+        // GET a single file within the bundle
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/support_bundle/storage.rs:982]: Sure thing, updated

File: sled-agent/src/support_bundle/storage.rs [hunk 19]
```diff
@@ -0,0 +1,1393 @@
...
+                response.headers()[CONTENT_LENGTH],
+                ((last + 1) - first).to_string(),
+            );
+            assert_eq!(
+                response.headers()[CONTENT_TYPE],
+                "application/octet-stream"
+            );
+            assert_eq!(response.headers()[ACCEPT_RANGES], "bytes");
+        }
+
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/support_bundle/storage.rs:1517]: Sure, can do

File: sled-agent/api/src/lib.rs (hunk 8 above)
Comment (smklein) [reply sled-agent/api/src/lib.rs]: Updated

**Comment by smklein** [comment]
Thanks for the feedback! All the responses to review comments should be contained within e1e185bbc3640cb99a1c0b9049ad520981b6b283

**Review by wfchandler — APPROVED** [review]
LGTM, thanks @smklein!
